qualtran/bloqs/arithmetic/addition.py

```diff
@@ -359,6 +361,63 @@ def _cvs_converter(vv):
     return tuple(int(v) for v in vv)
 
 
+@frozen
+class XorK(Bloq):
+    r"""Maps |x> to |x \oplus k> for a constant k.
+
+    Args:
+        dtype: Data type of the input register `x`.
+        k: The classical integer value to be XOR-ed to x.
+        cvs: A tuple of control values. Each entry specifies whether that control line is a
+            "positive" control (`cv[i]=1`) or a "negative" control (`cv[i]=0`).
+
+    Registers:
+        x: A quantum register of type `self.dtype` (see above).
+        ctrls: A sequence of control qubits (only when `cvs` is non-empty).
+    """
+    dtype: QDType
+    k: SymbolicInt
+    cvs: Tuple[int, ...] = field(converter=_cvs_converter, default=())
+
+    @cached_property
+    def signature(self) -> 'Signature':
+        return Signature(
+            ((Register('ctrls', QBit(), shape=(len(self.cvs),)),) if len(self.cvs) > 0 else ())
+            + (Register('x', self.dtype),)
+        )
+
+    @cached_property
+    def bitsize(self) -> SymbolicInt:
+        return self.dtype.num_qubits
+
+    def is_symbolic(self):
+        return is_symbolic(self.k, self.dtype)
+
+    def build_composite_bloq(self, bb: 'BloqBuilder', **soqs: 'SoquetT') -> Dict[str, 'SoquetT']:
+        if self.is_symbolic():
+            raise DecomposeTypeError(f"cannot decompose symbolic {self}")
+
+        xs = bb.split(cast(Soquet, soqs.pop('x')))
+        ctrls = soqs.pop('ctrls', None)
+
+        for i, bit in enumerate(self.dtype.to_bits(self.k)):
+            if bit == 1:
+                if len(self.cvs) > 0 and ctrls is not None:
+                    ctrls, xs[i] = bb.add(MultiControlX(cvs=self.cvs), ctrls=ctrls, x=xs[i])
+                else:
+                    xs[i] = bb.add(XGate(), q=xs[i])
+
+        soqs['x'] = bb.join(xs)
+        if ctrls is not None:
+            soqs['ctrls'] = ctrls
+        return soqs
+
+    def build_call_graph(self, ssa: 'SympySymbolAllocator') -> Set['BloqCountT']:
+        bit_flip_bloq = MultiControlX(cvs=self.cvs) if len(self.cvs) > 0 else XGate()
+        num_flips = self.bitsize if self.is_symbolic() else sum(self.dtype.to_bits(self.k))
+        return {(bit_flip_bloq, num_flips)}
```

Comment:
We shouldn't do `n` `MCX` gates to do `n` bit flips when number of controls > 1

Instead, divide it into 3 cases:

1) Num controls == 0 : Simply do an XGate()
2) Num controls == 1 : Simply do CNOT() gates
3) Num controls > 1 : Use a temporary ancilla and do a `MultiAnd(cvs=self.cvs)`  with the temporary ancilla as the target. Then reduce to Case(2) with the temporary ancilla as control.

Reply:
In fact, I'd suggest we only support Case (1) and (2) and reducing Case (3) to Case (1) is a general strategy used for all multi controlled bloq so we can add a separate Bloq that does this reduction (maybe at a later point)

Reply:
I simplified this to only support a single `control_val`, as #1131 will handle the CtrlSpec -> QBit part.